### PR TITLE
[Error categories] Support.

### PR DESCRIPTION
**Pull request by antmat**
Draft PR. Need some cleanup and tests.

### CODE REVIEW

File: unicorn/include/cocaine/unicorn/errors.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#include "zookeeper/zookeeper.h"
+
+#include <system_error>
+
+namespace cocaine { namespace error {
+
+typedef ZOO_ERRORS zookeeper_errors;
+
+/**
+* Extensions of zookeeper error used by cocaine. values greater than -1000 are reserved for zookeeper.
```
Comment (3Hren) [inline unicorn/include/cocaine/unicorn/errors.hpp]: This comment seems misleading.

File: unicorn/include/cocaine/zookeeper/zookeeper.hpp [hunk 2]
```diff
@@ -46,15 +30,15 @@ typedef long version_t;
 /**
 * Get zookeper error message by error code (including extra codes ZOO_EXTRA_ERROR)
 */
-std::string
-get_error_message(int rc);
+//std::string
```
Comment (3Hren) [inline unicorn/include/cocaine/zookeeper/zookeeper.hpp]: Dead code.

File: unicorn/include/cocaine/zookeeper/zookeeper.hpp [hunk 3]
```diff
@@ -46,15 +30,15 @@ typedef long version_t;
...
+//get_error_message(int rc);
 
 /**
 * Get zookeper error message by error code (including extra codes ZOO_EXTRA_ERROR)
 * and append exception message
 */
-std::string
-get_error_message(int rc, const std::exception& e);
+//std::string
```
Comment (3Hren) [inline unicorn/include/cocaine/zookeeper/zookeeper.hpp]: Ditto.

File: unicorn/src/cluster/unicorn.cpp [hunk 4]
```diff
@@ -86,13 +86,13 @@ unicorn_cluster_t::on_announce::write(unicorn::api_t::response::create_result&&
...
-unicorn_cluster_t::on_announce::abort(int rc, const std::string& reason) {
+unicorn_cluster_t::on_announce::abort(const std::error_code& rc) {
     //Ok for us.
-    if(rc == ZNODEEXISTS) {
+    if(rc.value() == ZNODEEXISTS) {
         COCAINE_LOG_INFO(parent->log, "announce checked");
         return;
     }
-    COCAINE_LOG_ERROR(parent->log, "could not announce local services: %s", reason.c_str());
+    COCAINE_LOG_ERROR(parent->log, "could not announce local services(%i): %s.", rc.value(), rc.message());
```
Comment (3Hren) [inline unicorn/src/cluster/unicorn.cpp]: Excess dot.

File: unicorn/src/cluster/unicorn.cpp [hunk 5]
```diff
@@ -186,8 +186,8 @@ unicorn_cluster_t::on_list_update::write(unicorn::api_t::response::children_subs
 }
 
 void
-unicorn_cluster_t::on_list_update::abort(int rc, const std::string& reason) {
-    COCAINE_LOG_WARNING(parent->log, "failure during subscription(%i): %s, resubscribing.", rc, reason.c_str());
+unicorn_cluster_t::on_list_update::abort(const std::error_code& rc) {
+    COCAINE_LOG_WARNING(parent->log, "failure during subscription(%i): %s, resubscribing.", rc.value(), rc.message().c_str());
```
Comment (3Hren) [inline unicorn/src/cluster/unicorn.cpp]: Dot.

File: unicorn/src/unicorn/errors.cpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            case INVALID_NODE_NAME:
+                return "Inavlid node name specified";
+            case INVALID_PATH:
+                return "Inavlid path specified";
+            case VERSION_NOT_ALLOWED:
+                return "Specified version is not allowed for command";
+            case INVALID_CONNECTION_ENDPOINT:
+                return "Invalid connection endpoint specified";
+            default:
+                return "Unknow unicorn error";
```
Comment (3Hren) [inline unicorn/src/unicorn/errors.cpp]: `"Unknow unicorn error " + code` would be more helpful.

**Comment by 3Hren** [comment]
:+1:

**Comment by noxiouz** [comment]
+1

**Pull request merged**